workers/unity/Packages/com.improbable.gdk.core/Systems/CleanReactiveComponentsSystem.cs

```diff
@@ -65,27 +58,27 @@ private void GenerateComponentGroups()
                     }
 
                     typesToRemove.Add(type);
-                    addRemoveComponentActionMethod.MakeGenericMethod(type).Invoke(this, null);
                 }
-            }
+            }            
         }
 
-        private void AddRemoveComponentAction<T>()
+        private void RemoveComponents()
         {
-            var componentGroup = GetComponentGroup(ComponentType.ReadOnly<T>());
-            removeComponentActions.Add(() =>
+            var componentToRemove = new List<Tuple<Entity,Type>>();
+            foreach (Type type in typesToRemove)
             {
-                if (componentGroup.IsEmptyIgnoreFilter)
-                {
-                    return;
-                }
-
+                var componentGroup = GetComponentGroup(ComponentType.ReadOnly(type));
                 var entityArray = componentGroup.GetEntityArray();
                 for (var i = 0; i < entityArray.Length; ++i)
                 {
-                    PostUpdateCommands.RemoveComponent<T>(entityArray[i]);
+                    componentToRemove.Add(Tuple.Create<Entity, Type>(entityArray[i], type));
                 }
-            });
+            }
+
+            foreach (Tuple<Entity, Type> entity in componentToRemove)
+            {
+                EntityManager.RemoveComponent(entity.Item1, entity.Item2);
```

Comment:
If I understood correctly, ComponentGroup is invalidated when components is removed.
This code grabs all the entities to the list first while all componentGroups are still valid and removes them one by one afterwards

Reply:
Yup that _should_ work as far as I understand it.

You can use C#7 tuples I think to cut down the verbosity: 
`var componentToRemove = new List<(Entity, Type)>();` 
and
`componentToRemove.Add((entityArray[i], type));` 

Will need to check the performance of this code however, cc @samiwh

Reply:
Checked this in profiler and it looks like CleanReactiveComponentSystem roughly takes the same time 2-3ms